### ISSUE TITLE
moonshine v2: ForwardDecoder feeds encoder_attention_mask at frame length; graphs expect raw audio length (silent empty decode above 384 encoder frames)

### ISSUE THREAD

**Issue by paoloantinori** [k2-fsa/sherpa-onnx#3975]
## Summary

`OfflineMoonshineModelV2::Impl::ForwardDecoder` feeds the decoder graph's `encoder_attention_mask` at encoder-FRAME length, but the moonshine v2 decoder graphs expect it at RAW audio length (the transformers 4.49 contract) and downsample it internally. Above 384 encoder frames the graph throws an onnxruntime broadcast error, and `DecodeStream` catches it and returns an empty result, so long audio silently transcribes to nothing.

## What happens

sherpa-onnx 1.13.8 (python wheel, CPU). The mask-bearing moonshine v2 exports from the 2026-02-27 line (I tested `csukuangfj2/sherpa-onnx-base-{uk,ar,vi}-quantized-2026-02-27`, i.e. the `sherpa-onnx-moonshine-*` repos) decode correctly up to 384 encoder frames and return empty text above that. The log shows only:

```
Caught exception:
Non-zero status code returned while running Add node. Name:'/model/decoder/layers.0/encoder_attn/Add' Status Message: ... Attempting to broadcast an axis by a dimension other than 1. 2 by 403
Return an empty result. Number of audio samples: 155520
```

## Why

In `sherpa-onnx/csrc/offline-moonshine-model-v2.cc` (1.13.8), `ForwardDecoder` builds the mask from the encoder OUTPUT length:

```cpp
mask.resize(encoder_out.GetTensorTypeAndShapeInfo().GetShape()[1], 1);
```

`ForwardEncoder` in the same file correctly feeds the mask at the raw audio length. The decoder graphs follow the transformers 4.49 moonshine modeling, where the decoder downsamples the mask itself:

```python
downsample_stride = 64 * 3 * 2  # 384, the conv strides
encoder_attention_mask = encoder_attention_mask[..., ::downsample_stride][..., :mask_len]
```

With a frame-length mask of N, the `::384` slice yields `ceil(N/384)` elements: 1 for N <= 384 (broadcasts, works by accident), 2 for 385..768 (the `2 by 403` above, 2 = ceil(403/384), computed at runtime), and so on for longer inputs. The graphs themselves are fine: driving the decoder directly in onnxruntime, the same files run clean at 385, 403 and 800 encoder frames when fed a raw-length mask.

Reproducer (no sherpa needed):

```python
import numpy as np, onnxruntime as ort, re

def drive(model_path, enc_len, raw_mask):
    so = ort.SessionOptions(); so.log_severity_level = 3
    s = ort.InferenceSession(model_path, so, providers=["CPUExecutionProvider"])
    inputs = {i.name: i.shape for i in s.get_inputs()}
    dmodel = inputs["encoder_hidden_states"][2]
    layers = max(int(m.group(1)) for n in inputs
                 if (m := re.match(r"past_key_values\.(\d+)\.", n)))
    nh, hd = next((sh[1], sh[3]) for n, sh in inputs.items()
                  if n.startswith("past_key_values.0.decoder.key"))
    enc = (np.random.default_rng(0).standard_normal((1, enc_len, dmodel)) * 0.1).astype(np.float32)
    feed = {"input_ids": np.array([[1]], dtype=np.int64),
            "encoder_hidden_states": enc,
            "use_cache_branch": np.array([False], dtype=bool)}
    if "encoder_attention_mask" in inputs:
        n = enc_len * 384 if raw_mask else enc_len
        feed["encoder_attention_mask"] = np.ones((1, n), dtype=np.int64)
    for i in range(layers + 1):
        for w in ("decoder.key", "decoder.value", "encoder.key", "encoder.value"):
            feed[f"past_key_values.{i}.{w}"] = np.zeros((1, nh, 0, hd), dtype=np.float32)
    s.run(None, feed)

# decoder_model_merged.ort of sherpa-onnx-moonshine-base-uk-quantized-2026-02-27:
#   drive(p, 384, raw_mask=False) -> runs (mask collapses to 1 element)
#   drive(p, 385, raw_mask=False) -> broadcast error "2 by 385"
#   drive(p, 403, raw_mask=False) -> broadcast error "2 by 403"
#   drive(p, 385, raw_mask=True)  -> runs clean (and 403, 800 too)
```

The `es` export of the same line is unaffected because its decoder has no `encoder_attention_mask` input at all (the `decoder_needs_mask_` flag path), which is what made the failure look model-specific.

## Suggested fix

Feed all-ones at raw length in `ForwardDecoder`, exactly as `ForwardEncoder` already does. Any all-ones length >= (frames - 1) * 384 + 1 is semantically correct for unmasked audio, so `frames * 384` is a safe choice if the true raw length is not at hand there.

## Secondary ask

`OfflineRecognizerMoonshineV2Impl::DecodeStream` catches `Ort::Exception` and returns an empty result. That design hid this completely on our side (an Android app chunking moonshine at 30s shipped and the only symptom was missing transcript text). Would you consider surfacing decode exceptions, or at least returning a distinguishable error, instead of an empty success?

**Comment by csukuangfj** [comment]
Is it the same issue as #3471

**Comment by paoloantinori** [comment]
Same defect, yes: the signature matches (the `2 by 415` there, `2 by 403` here; both are `ceil(frames/384)`).

I would revisit the conclusion from #3471, though. The shipped graphs are not hardcoded-broken: driven directly in onnxruntime, the same `decoder_model_merged.ort` runs clean at 385/403/800 encoder frames when `encoder_attention_mask` is fed at raw audio length, and fails only with the frame-length mask `ForwardDecoder` builds. The graph downsamples the mask itself (`mask[..., ::384]`, the conv strides), so a frame-length mask collapses to `ceil(frames/384)` elements; that is where the 2 comes from. `ForwardEncoder` already feeds the raw length for the encoder mask, so the fix is the symmetric one-liner in `ForwardDecoder`. I am building a branch with the change now and will open the PR once I have decoded long files with it.

Full analysis and the pure-ORT reproducer (both mask lengths, the boundary table): https://github.com/RisorseArtificiali/anti-vocale/blob/main/docs/research/2026-09-22-moonshine-v2-384-frame-limit.md

**Comment by paoloantinori** [comment]
PR with the fix: #3976 (verified by decoding a 20.4s clip that returns empty on 1.13.8).
